Automattic/jetpack
# Jetpack_Gutenberg: Simplify

Refactor `Jetpack_Gutenberg` to get rid of some of its quirks, and make it easier to reason about. Among other things, this allows us to remove the Mailchimp workaround (https://github.com/Automattic/jetpack/pull/10597/files#r245018740) as part of this PR (per 293b373109f49231db13f1c726298d7655115c99).

Some basic ideas:

- Decouple registration, and storing of availability information. Only when a block isn't being registered do we set the latter.
- No need for a duplicated block registry -- simply use Gutenberg's.
- Fewer getters/setters in `Jetpack_Gutenberg`, thus fewer (potentially broken) states. The idea is that `Jetpack_Gutenberg` initializes a whitelist of potentially allowed blocks and plugins once, and upon (attempted) registration of individual blocks and plugins, checks if they're on that list. This is probably best reflected by the changes to the unit test.
- Remove layers of indirection, especially wrt filters. Instead of passing a callback when setting availability in more complex scenarios, we simply move block registration and availability setting to an appropriate filter _inside_ the corresponding module. This should make the order of actions and flow of filters much easier to reason about.
- Removal of asymmetry (there were some methods for plugins that didn't actually do anything).
- Filter child blocks from availability tracking so we don't confuse them with other unwhitelisted blocks.

#### Changes proposed in this Pull Request:

- [x] Introduce and use set_extension_unavailability_reason() (to decouple block registration and availability)
- [x] Update endpoint to directly fetch registration information from GB block registry
- [x] Remove obsolete stuff from `Jetpack_Gutenberg`
- [x] Rebase, remove registration workarounds from #10597
- [x] Fix (and update) tests
- [x] Update comments

#### Testing instructions:

Verify that block availability works as before. Verify that the following extensions's availability changes when doing the following:
- Publicize:
  - Toggle the Publicize module, verify that the Publicize panel only shows when the module is on
  - With a user with 'contributor' role, write a new post, and verify that the Publicize panel doesn't show in the Jetpack sidebar. (If you check `window.Jetpack_Editor_Initial_State`, it should say `unavailable_reason: "unauthorized"` for `publicize`.)
- Simple Payments: Try a free vs a premium plan
- Contact Form:
  - Start writing a post, and insert a Contact Form Block
  - Modify some of the Contact Form's field labels (e.g. change 'Name' to 'Your Name')
  - Publish the post, and view it in the frontend.
  - Verify that your modifications actually show up.

Change the `JETPACK_BETA_BLOCKS` constant and verify that depending on its value, beta blocks are either available or not!

#### Proposed changelog entry for your changes:

* Simplify the `Jetpack_Gutenberg` class
